### PR TITLE
move fetchBlocks to node-core

### PR DESCRIPTION
**Pull request by DeveloperInProgress**
# Description
move fetchBlocks of ApiService to node-core to improve code reuse

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch

### CODE REVIEW

**Review by jiqiang90 — APPROVED** [review]

File: packages/node-core/src/api.service.ts [hunk 1]
```diff
@@ -13,5 +21,25 @@ export abstract class ApiService {
 
   abstract init(): Promise<ApiService>;
 
+  async fetchBlocksGeneric<T>(
+    batch: number[],
+    fetchFuncProvider: FetchFunctionProvider<T>,
+    overallSpecVer?: number
```
Comment (stwiname) [inline packages/node-core/src/api.service.ts:27]: Spec version is specific to substrate

File: packages/node-core/src/api.service.ts [hunk 2]
```diff
@@ -13,5 +21,25 @@ export abstract class ApiService {
...
+        return blocks;
+      } catch (e) {
+        logger.error(e, `Failed to fetch blocks ${batch[0]}...${batch[batch.length - 1]}`);
+
+        reconnectAttempts++;
+      }
+    }
+    throw new Error(`Maximum number of retries (${MAX_RECONNECT_ATTEMPTS}) reached.`);
+  }
   abstract get api(): any; /*ApiWrapper*/
```
Comment (stwiname) [inline packages/node-core/src/api.service.ts:44]: Can we keep the abstract methods together please

File: packages/node-core/src/api.service.ts [hunk 3]
```diff
@@ -13,5 +21,25 @@ export abstract class ApiService {
...
+    fetchFuncProvider: FetchFunctionProvider<T>,
+    overallSpecVer?: number
+  ): Promise<T[]> {
+    let reconnectAttempts = 0;
+    while (reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
+      try {
+        // Get the latest fetch function from the provider
+        const fetchFunc = fetchFuncProvider();
+        const blocks = await fetchFunc(batch, overallSpecVer);
+        return blocks;
```
Comment (stwiname) [inline packages/node-core/src/api.service.ts:35]:
```suggestion
        return await fetchFunc(batch, overallSpecVer);
```

File: packages/node/src/indexer/api.service.ts [hunk 4]
```diff
@@ -251,54 +257,15 @@ export class ApiService implements OnApplicationShutdown {
...
-        reconnectAttempts++;
-      }
-    }
-    throw new Error(
-      `Maximum number of retries (${MAX_RECONNECT_ATTEMPTS}) reached.`,
+    return this.fetchBlocksGeneric<BlockContent>(
+      batch,
+      () => (blockArray: number[], specVer?: number) =>
+        this.fetchBlocksBatches(this.api, blockArray, specVer),
+      [overallSpecVer],
```
Comment (stwiname) [inline packages/node/src/indexer/api.service.ts:268]:
```suggestion
    return this.fetchBlocksGeneric<BlockContent>(
      batch,
      () => (blockArray: number[]) =>
        this.fetchBlocksBatches(this.api, blockArray, opverallSpecVer),
```

File: packages/node-core/src/api.service.ts [hunk 5]
```diff
@@ -12,6 +20,21 @@ export abstract class ApiService {
   constructor(protected project: any) {}
 
   abstract init(): Promise<ApiService>;
-
   abstract get api(): any; /*ApiWrapper*/
+
+  async fetchBlocksGeneric<T>(batch: number[], fetchFuncProvider: FetchFunctionProvider<T>, args?: any): Promise<T[]> {
```
Comment (stwiname) [inline packages/node-core/src/api.service.ts:25]: We don't need args and we should make attempts configurabel

```suggestion
  async fetchBlocksGeneric<T>(fetchFuncProvider: FetchFunctionProvider<T>, batch: number[], numAttempts = MAX_RECONNECT_ATTEMPTS): Promise<T[]> {
```

**Review by stwiname — APPROVED** [review]

**Pull request merged**